### ISSUE TITLE
"@" was removed in the header line of output fastq file 

### ISSUE THREAD

**Issue by hongenxu** [biocore/sortmerna#131]
Hi, 

I used the following command

```
${sortmerna}/sortmerna -a 8 --ref \
 ${sortmerna}/rRNA_databases/silva-bac-16s-id90.fasta,${sortmerna}/index/silva-bac-16s-db:${sortmerna}/rRNA_databases/silva-bac-23s-id98.fasta,${sortmerna}/index/silva-bac-23s-db:${sortmerna}/rRNA_databases/silva-arc-16s-id95.fasta,${sortmerna}/index/silva-arc-16s-db:${sortmerna}/rRNA_databases/silva-arc-23s-id98.fasta,${sortmerna}/index/silva-arc-23s-db:${sortmerna}/rRNA_databases/silva-euk-18s-id95.fasta,${sortmerna}/index/silva-euk-18s-db:${sortmerna}/rRNA_databases/silva-euk-28s-id98.fasta,${sortmerna}/index/silva-euk-28s:${sortmerna}/rRNA_databases/rfam-5s-database-id98.fasta,${sortmerna}/index/rfam-5s-db:${sortmerna}/rRNA_databases/rfam-5.8s-database-id98.fasta,${sortmerna}/index/rfam-5.8s-db \
          --reads-gz ${output_dir}/${sample}.fq.gz \
          --other ${output_dir}/${sample}.other --aligned ${output_dir}/${sample}.align\
          --log -v --paired_in --fastx

```
In the input file, the first 10 lines are 
@HWI-D00731:62:C6F5EANXX:5:1101:2305:1937 1:N:0:GAGTCA
NGGGTATGTGTTAATATACTTGAAAGCATGAACAGAACATATACGCACAGGAAGTGAACTTGAAGGTCTAGAGAGTTCTTCCACTCTGCCTTTTTTGATTTTTTTCTCCTTTGTTTTAACACTTG
+
#=<ABGGGGGGGGGGGGGGGGGGEGEBGGGGGGGGEGGGGGGGGGGGGGGGGGGGGCGGGGGGGGGFGGGGGGGGCFGGGGEGGGGCFGGC0CGGFGGGFGG0:CDGG>B0FCGBGGGGG@F@@F
@HWI-D00731:62:C6F5EANXX:5:1101:2305:1937 2:N:0:GAGTCA
CTGCAGGGTAAATGCAGACAGAAAACTGAAGCAAGGAAGAACAGGTCCCTCACCATTTCGTGGGTGGTGGCACTATTTAAAAGGGCACTAATGTCACAATGTAATGATTCACAAGTGTTAAAACA
+
CCCCCGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGFGGGGGGGGGGGGGGGGGGGGGGGGGGGGGEGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGEGGGGGGGGEGG
@HWI-D00731:62:C6F5EANXX:5:1101:2429:1986 1:N:0:GAGTCA
GTGACTTATGCTTACTGGAAGCTGATTAGTGTAAGGTGTAAGTTGATATCTTTCTACTTTTAGTGCACTGATGTGATCTGGAAAAATCCATTTTTACCAAAATAGGAATACTTAAGTCATTCAAA

but in the output "other" and "aligned" files, 
the first 10 lines are 

HWI-D00731:62:C6F5EANXX:5:1101:2602:1923 1:N:0:GAGTCA
NAGATAATACATGCCGACGAGCGCCGACCTCCGGGGACGCGTGCATTTATCAGACCAAAACCAACCCGGGCTCGCCCGGCGGCTTTGGTGACTCTAGATAACCTCGAGCCGATCGCACGCCCCCG
+
#<=0=E@FD@GFF1DB>>G<CAGDAGDGEEGEGGGGGADGGGGGBDGCDGCGEC@00<<CGGG=G<>EGGGGGGA9CDGGGGGCGGEGDEC/7C@/@8@D/:DD<8.66.C>D6>..69C..68C
HWI-D00731:62:C6F5EANXX:5:1101:2602:1923 2:N:0:GAGTCA
ACCCGTGGTCACCATGGTAGGCACAGACAGTACCATCGAAAGTTGATAGGGCAGACATTCGAATGGGTCGTCGCCGCCACGGGGGCGTGCGATCGGCTCGAGGTTATCTAGAGTCACCAAAGCCG
+
<<AA<?FGF1FGCGGGGGFGGCGG>E>FDGG@1E>F@GEFGEEFGE@GGGG1===11EFGG<>FFGGF>EGGGGGGGGDGGGGGGAG6EBGBBGGGDGCC8.C9.CGG/DC6:6C/C/CC8.:6C
HWI-D00731:62:C6F5EANXX:5:1101:2787:1898 1:N:0:GAGTCA
NCAGCTATCAACTTTCGATGGTACTGTCTGTGCCTACCATGGTGACCACGGGTAACGGGGAATCAGGGTTCGATTCCGGAGAGGGAGCCTGAGAAACGGCTACCACATCCAAGGAAGGCAGCAGG

So, the "@" was missing in the output files.

**Comment by ekopylova** [comment]
How many sections was your input file processed in (it should say in the log file)?

Thanks!

**Comment by hongenxu** [comment]
Sorry, I cannot find the "sections" in the log file and align.log. 

I listed both log files. 

**align.log**
 Wed Dec 14 11:12:07 2016

 Command: /home/users/xu/sortmerna/sortmerna -a 8 --ref /home/users/xu/sortmerna/rRNA_databases/silva-bac-16s-id90.fasta,/home/users/xu/sortmerna/index/silva-bac-16s-db:/home/users/xu/sortmerna/rRNA_databases/silva-bac-23s-id98.fasta,/home/users/xu/sortmerna/index/silva-bac-23s-db:/home/users/xu/sortmerna/rRNA_databases/silva-arc-16s-id95.fasta,/home/users/xu/sortmerna/index/silva-arc-16s-db:/home/users/xu/sortmerna/rRNA_databases/silva-arc-23s-id98.fasta,/home/users/xu/sortmerna/index/silva-arc-23s-db:/home/users/xu/sortmerna/rRNA_databases/silva-euk-18s-id95.fasta,/home/users/xu/sortmerna/index/silva-euk-18s-db:/home/users/xu/sortmerna/rRNA_databases/silva-euk-28s-id98.fasta,/home/users/xu/sortmerna/index/silva-euk-28s:/home/users/xu/sortmerna/rRNA_databases/rfam-5s-database-id98.fasta,/home/users/xu/sortmerna/index/rfam-5s-db:/home/users/xu/sortmerna/rRNA_databases/rfam-5.8s-database-id98.fasta,/home/users/xu/sortmerna/index/rfam-5.8s-db --reads-gz /scratch/xu/MDV_project/deg/analysis/rRNA/test.fq.gz --other /scratch/xu/MDV_project/deg/analysis/rRNA/test.other.fastq fastq /scratch/xu/MDV_project/deg/analysis/rRNA/test.align --log -v --paired_in --fastx
 Process pid = 3806
 Parameters summary:
    Index: /home/users/xu/sortmerna/index/silva-bac-16s-db
     Seed length = 18
     Pass 1 = 18, Pass 2 = 9, Pass 3 = 3
     Gumbel lambda = 0.602725
     Gumbel K = 0.329559
     Minimal SW score based on E-value = 48
    Index: /home/users/xu/sortmerna/index/silva-bac-23s-db
     Seed length = 18
     Pass 1 = 18, Pass 2 = 9, Pass 3 = 3
     Gumbel lambda = 0.602436
     Gumbel K = 0.335011
     Minimal SW score based on E-value = 48
    Index: /home/users/xu/sortmerna/index/silva-arc-16s-db
     Seed length = 18
     Pass 1 = 18, Pass 2 = 9, Pass 3 = 3
     Gumbel lambda = 0.596286
     Gumbel K = 0.322254
     Minimal SW score based on E-value = 46
    Index: /home/users/xu/sortmerna/index/silva-arc-23s-db
     Seed length = 18
     Pass 1 = 18, Pass 2 = 9, Pass 3 = 3
     Gumbel lambda = 0.597507
     Gumbel K = 0.331576
     Minimal SW score based on E-value = 44
    Index: /home/users/xu/sortmerna/index/silva-euk-18s-db
     Seed length = 18
     Pass 1 = 18, Pass 2 = 9, Pass 3 = 3
     Gumbel lambda = 0.612551
     Gumbel K = 0.339810
     Minimal SW score based on E-value = 47
    Index: /home/users/xu/sortmerna/index/silva-euk-28s
     Seed length = 18
     Pass 1 = 18, Pass 2 = 9, Pass 3 = 3
     Gumbel lambda = 0.612082
     Gumbel K = 0.345772
     Minimal SW score based on E-value = 47
    Index: /home/users/xu/sortmerna/index/rfam-5s-db
     Seed length = 18
     Pass 1 = 18, Pass 2 = 9, Pass 3 = 3
     Gumbel lambda = 0.616694
     Gumbel K = 0.342032
     Minimal SW score based on E-value = 46
    Index: /home/users/xu/sortmerna/index/rfam-5.8s-db
     Seed length = 18
     Pass 1 = 18, Pass 2 = 9, Pass 3 = 3
     Gumbel lambda = 0.617555
     Gumbel K = 0.343861
     Minimal SW score based on E-value = 44
    Number of seeds = 2
    Edges = 4 (as integer)
    SW match = 2
    SW mismatch = -3
    SW gap open penalty = 5
    SW gap extend penalty = 2
    SW ambiguous nucleotide = -3
    SQ tags are not output
    Number of threads = 8
    Reads file = /scratch/xu/MDV_project/deg/analysis/rRNA/test.fq.gz

 Results:
    Total reads = 10000
    Total reads passing E-value threshold = 2443 (24.43%)
    Total reads failing E-value threshold = 7557 (75.57%)
    Minimum read length = 125
    Maximum read length = 125
    Mean read length = 125
 By database:
    /home/users/xu/sortmerna/rRNA_databases/silva-bac-16s-id90.fasta            0.55%
    /home/users/xu/sortmerna/rRNA_databases/silva-bac-23s-id98.fasta            1.13%
    /home/users/xu/sortmerna/rRNA_databases/silva-arc-16s-id95.fasta            0.05%
    /home/users/xu/sortmerna/rRNA_databases/silva-arc-23s-id98.fasta            0.01%
    /home/users/xu/sortmerna/rRNA_databases/silva-euk-18s-id95.fasta            16.09%
    /home/users/xu/sortmerna/rRNA_databases/silva-euk-28s-id98.fasta            6.58%
    /home/users/xu/sortmerna/rRNA_databases/rfam-5s-database-id98.fasta         0.02%
    /home/users/xu/sortmerna/rRNA_databases/rfam-5.8s-database-id98.fasta               0.00%


**logfile**

  Program:     SortMeRNA version 2.1b, 03/03/2016
  Copyright:    2012-17 Bonsai Bioinformatics Research Group:
                LIFL, University Lille 1, CNRS UMR 8022, INRIA Nord-Europe
                2014-17 Knight Lab:
                Department of Pediatrics, UCSD, La Jolla
  Disclaimer:   SortMeRNA comes with ABSOLUTELY NO WARRANTY; without even the
                implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
                See the GNU Lesser General Public License for more details.
  Contributors: Jenya Kopylova   jenya.kopylov@gmail.com
                Laurent Noé      laurent.noe@lifl.fr
                Pierre Pericard  pierre.pericard@lifl.fr
                Daniel McDonald  wasade@gmail.com
                Mikaël Salson    mikael.salson@lifl.fr
                Hélène Touzet    helene.touzet@lifl.fr
                Rob Knight       robknight@ucsd.edu


  Computing read file statistics ...  done [0.03 sec]
  Parameters summary:
    Number of seeds = 2
    Edges = 4 (as integer)
    SW match = 2
    SW mismatch = -3
    SW gap open penalty = 5
    SW gap extend penalty = 2
    SW ambiguous nucleotide = -3
    SQ tags are not output
    Number of threads = 8
  Time to load reads and set up pointers [0.02 sec]

  Begin analysis of: /home/users/xu/sortmerna/rRNA_databases/silva-bac-16s-id90.fasta
    Seed length = 18
    Pass 1 = 18, Pass 2 = 9, Pass 3 = 3
    Gumbel lambda = 0.602725
    Gumbel K = 0.329559
    Minimal SW score based on E-value = 48
    Loading index part 1/1 ...  done [3.13 sec]
    Begin index search ...  done [0.63 sec]
    Freeing index ...  done [0.53 sec]

  Begin analysis of: /home/users/xu/sortmerna/rRNA_databases/silva-bac-23s-id98.fasta
    Seed length = 18
    Pass 1 = 18, Pass 2 = 9, Pass 3 = 3
    Gumbel lambda = 0.602436
    Gumbel K = 0.335011
    Minimal SW score based on E-value = 48
    Loading index part 1/1 ...  done [3.31 sec]
    Begin index search ...  done [0.21 sec]
    Freeing index ...  done [0.39 sec]

  Begin analysis of: /home/users/xu/sortmerna/rRNA_databases/silva-arc-16s-id95.fasta
    Seed length = 18
    Pass 1 = 18, Pass 2 = 9, Pass 3 = 3
    Gumbel lambda = 0.596286
    Gumbel K = 0.322254
    Minimal SW score based on E-value = 46
    Loading index part 1/1 ...  done [1.43 sec]
    Begin index search ...  done [0.18 sec]
    Freeing index ...  done [0.11 sec]

  Begin analysis of: /home/users/xu/sortmerna/rRNA_databases/silva-arc-23s-id98.fasta
    Seed length = 18
    Pass 1 = 18, Pass 2 = 9, Pass 3 = 3
    Gumbel lambda = 0.597507
    Gumbel K = 0.331576
    Minimal SW score based on E-value = 44
    Loading index part 1/1 ...  done [0.61 sec]
    Begin index search ...  done [0.12 sec]
    Freeing index ...  done [0.06 sec]

  Begin analysis of: /home/users/xu/sortmerna/rRNA_databases/silva-euk-18s-id95.fasta
    Seed length = 18
    Pass 1 = 18, Pass 2 = 9, Pass 3 = 3
    Gumbel lambda = 0.612551
    Gumbel K = 0.339810
    Minimal SW score based on E-value = 47
    Loading index part 1/1 ...  done [3.00 sec]
    Begin index search ...  done [1.13 sec]
    Freeing index ...  done [0.38 sec]

  Begin analysis of: /home/users/xu/sortmerna/rRNA_databases/silva-euk-28s-id98.fasta
    Seed length = 18
    Pass 1 = 18, Pass 2 = 9, Pass 3 = 3
    Gumbel lambda = 0.612082
    Gumbel K = 0.345772
    Minimal SW score based on E-value = 47
    Loading index part 1/1 ...  done [3.27 sec]
    Begin index search ...  done [0.68 sec]
    Freeing index ...  done [0.44 sec]

  Begin analysis of: /home/users/xu/sortmerna/rRNA_databases/rfam-5s-database-id98.fasta
    Seed length = 18
    Pass 1 = 18, Pass 2 = 9, Pass 3 = 3
    Gumbel lambda = 0.616694
    Gumbel K = 0.342032
    Minimal SW score based on E-value = 46
    Loading index part 1/1 ...  done [1.75 sec]
    Begin index search ...  done [0.15 sec]
    Freeing index ...  done [0.18 sec]

  Begin analysis of: /home/users/xu/sortmerna/rRNA_databases/rfam-5.8s-database-id98.fasta
    Seed length = 18
    Pass 1 = 18, Pass 2 = 9, Pass 3 = 3
    Gumbel lambda = 0.617555
    Gumbel K = 0.343861
    Minimal SW score based on E-value = 44
    Loading index part 1/1 ...  done [0.68 sec]
    Begin index search ...  done [0.21 sec]
    Freeing index ...  done [0.07 sec]
    Total number of reads mapped (incl. all reads file sections searched): 2443
    Writing aligned FASTA/FASTQ ...  done [0.04 sec]
    Writing not-aligned FASTA/FASTQ ...  done [0.10 sec]

**Comment by lfaller** [comment]
I have also observed this. However I have only noticed it since I've been using the `--paired_in` option.

**Comment by lfaller** [comment]
Actually, I just viewed the output file with `cat -A` to show invisible characters, and I observed a caret character at the beginning of the line, just before the `@` symbol.

So I am wondering if sortmerna is trying to write the `@` character but something else gets mixed in that turns this into an invisible character?

```
cat -A Sample_2709359/sortmerna_out/2709359_rrna_free_reads_unmerged_R1.fastq | head
^@NB501485:33:HLLFTBGX2:1:11101:25825:1054 1:N:0:GCCTCGAAGTCC$
GCAGANTCGGCATACCATGAGGTGGGGGGNNNNGTGGGAGTANCNCNANATGTATGANTTTNTNNCTNANACACC$
+$
AAAAA#EEEEE/AEEEE/EEEEEEEAEEE####E/EEA//A/#A#E#<#EE//E///#EEE#/##//#E#AA/E<$
^@NB501485:33:HLLFTBGX2:1:11101:18173:1062 1:N:0:GCCTCGAAGTCC$
CCTTGATGTAAGGATCAGAGGCATCGTCCTCTGCATCGAACATGGCAGCCAGAAGATGGATTGGCTGGAGAGTGCC$
+$
AAAAAEEEE6EEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEA$
^@NB501485:33:HLLFTBGX2:1:11101:5476:1067 1:N:0:GCCTCGAAGTCC$
CTAATGAGTCAGTCGAGCAACAAAGTTGGCGCCAATGGCCAGGGCCAAGACAATACAAATGGCAGTGGGGCCGATC$
```

**Comment by lfaller** [comment]
If this is indeed the case, do you have a suggestion for how to replace this special character with the regular `@` symbol?

**Comment by ekopylova** [comment]
Hi there,
Could you try using the development version of SortMeRNA? The behavior you're seeing is not expected and I believe it could have been fixed in the dev version due to the approach reads are being input & output. Please refer to the [installation instructions](https://github.com/biocore/sortmerna#compilation) for the dev version and let me know if the problem persists there. I should be releasing a new version shortly.

**Comment by lfaller** [comment]
Actually I am using the dev version. I first downloaded the tar.gz binary file but couldn't get it to run with multiple cores. So then I tried the dev version as described on github and now I can use multiple cores but that "@" symbol issue cropped up.

**Comment by lfaller** [comment]
By the way, I have `sortmerna` installed on two different machines. On one machine, I installed the dev version as I described above and that's where I observed the malformed `@`. On another machine I installed the binary version of the tool and I did not observe the malformed `@`. On both of these machines I have recently run `sortmerna` with the `--paired_in` option. The help menu for the installation of `sortmerna` where I did not observe the malformed `@` symbol is as follows:

```
$ /home/lina/sortmerna/sortmerna

  Program:     SortMeRNA version 2.1, 01/02/2016
  Copyright:   2012-16 Bonsai Bioinformatics Research Group:
               LIFL, University Lille 1, CNRS UMR 8022, INRIA Nord-Europe
               2014-16 Knight Lab, Department of Pediatrics, UCSD, La Jolla,
  Disclaimer:  SortMeRNA comes with ABSOLUTELY NO WARRANTY; without even the
               implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
               See the GNU Lesser General Public License for more details.
  Contact:     Evguenia Kopylova, jenya.kopylov@gmail.com
               Laurent Noé, laurent.noe@lifl.fr
               Hélène Touzet, helene.touzet@lifl.fr

  For help or more information on usage, type `./sortmerna -h'
```

**Comment by lfaller** [comment]
@hongenxu if you can verify that you also see the invisible characters, I found this workaround to fix the PE read output files:

```
tr '\0' '@' < R1.fastq > R1_fixed.fastq
```

**Comment by ekopylova** [comment]
Is it possible to provide me with a minimal FASTQ file that shows this behavior? If I can reproduce the behavior, I will be able to debug quickly in the next few days. You can send me the file at jenya.kopylov @ gmail dot com.

**Comment by lfaller** [comment]
I made a small example and uploaded it to dropbox here: https://www.dropbox.com/sh/eptp3n2ukg4kd2k/AAC0_7-_i8-ilfUFxoMdekyVa?dl=0

Commands used to run this:

```
INPUT=small_example/small_example_merged.fastq
ALIGNED=small_example/small_example_rrna_reads
OTHER=small_example/small_example_rrna_free_reads

/home/ubuntu/sortmerna/sortmerna --ref /home/ubuntu/sortmerna/rRNA_databases/silva-bac-16s-id90.fasta,/home/ubuntu/sortmerna/index/silva-bac-16s-db:/home/ubuntu/sortmerna/rRNA_databases/silva-bac-23s-id98.fasta,/home/ubuntu/sortmerna/index/silva-bac-23s-db:/home/ubuntu/sortmerna/rRNA_databases/silva-arc-16s-id95.fasta,/home/ubuntu/sortmerna/index/silva-arc-16s-db:/home/ubuntu/sortmerna/rRNA_databases/silva-arc-23s-id98.fasta,/home/ubuntu/sortmerna/index/silva-arc-23s-db:/home/ubuntu/sortmerna/rRNA_databases/silva-euk-18s-id95.fasta,/home/ubuntu/sortmerna/index/silva-euk-18s-db:/home/ubuntu/sortmerna/rRNA_databases/silva-euk-28s-id98.fasta,/home/ubuntu/sortmerna/index/silva-euk-28s:/home/ubuntu/sortmerna/rRNA_databases/rfam-5s-database-id98.fasta,/home/ubuntu/sortmerna/index/rfam-5s-db:/home/ubuntu/sortmerna/rRNA_databases/rfam-5.8s-database-id98.fasta,/home/ubuntu/sortmerna/index/rfam-5.8s-db --reads $INPUT --num_alignments 1 --fastx --aligned $ALIGNED --other $OTHER --paired_in --log -v --blast 1
```

I hope this is helpful!

**Comment by stevendbrown** [comment]
I have this same problem with HEAD pointing to the current commit in master (d5b19dc).

**Comment by lfaller** [comment]
FYI, I am still observing this with sortmerna version 2.1b (the precompiled tool I downloaded from your releases page).

Thanks!

**Comment by ekopylova** [comment]
We are looking at this issue now, will report back as soon as it is resolved. Thanks for your patience :)

**Comment by lfaller** [comment]
Thank you for looking at it! 👍

**Comment by biocodz** [comment]
Applied fix (pull request 150). Tested on Ubuntu and Windows - OK

**Comment by ekopylova** [comment]
Fixed in PR #150
